lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php

```diff
@@ -87,4 +223,129 @@ protected static function get_blocks_metadata() {
+						foreach( $element['selectors'] as $sub_selector => $property ) {
+							$additional_rules .= $selector . $sub_selector . " { \n";
+							$additional_rules .= "\t" . $property . ': ' . $element['value'] . ";\n";
+							$additional_rules .= "}\n";
+						}
+						return $carry;
+					}
+					return $carry .= "\t" . $element['name'] . ': ' . $element['value'] . ";\n"; },
+				''
+			);
+
+			if ( $declaration_block ) {
+				$ruleset .= $selector . " {\n" . $declaration_block . "}\n";
+			}
+		} else {
+			$declaration_block = array_reduce(
+				$declarations,
+				function ( $carry, $element ) use( $selector, &$additional_rules ) {
+					if ( ! empty( $element['selectors'] ) ) {
+						foreach( $element['selectors'] as $sub_selector => $property ) {
+							$additional_rules .= $selector . $sub_selector . " { ";
+							$additional_rules .= $property . ': ' . $element['value'] . ";";
+							$additional_rules .= " }";
+						}
+						return $carry;
+					}
+					return $carry .= $element['name'] . ': ' . $element['value'] . ';'; },
+				''
+			);
+
+			if ( $declaration_block ) {
+				$ruleset .= $selector . '{' . $declaration_block . '}';
+			}
+		}
+
+		$ruleset .= $additional_rules;
+
+		return $ruleset;
+	}
+
+	/**
+	 * Given a styles array, it extracts the style properties
+	 * and adds them to the $declarations array following the format:
+	 *
+	 * ```php
+	 * array(
+	 *   'name'  => 'property_name',
+	 *   'value' => 'property_value,
+	 * )
+	 * ```
+	 *
+	 * @param array $styles Styles to process.
+	 * @param array $settings Theme settings.
+	 * @param array $properties Properties metadata.
+	 * @return array Returns the modified $declarations.
+	 */
+	protected static function compute_style_properties( $styles, $settings = array(), $properties = null ) {
+		if ( null === $properties ) {
+			$properties = static::PROPERTIES_METADATA;
+		}
+
+		$declarations = array();
+		if ( empty( $styles ) ) {
+			return $declarations;
+		}
+
+		foreach ( $properties as $css_property => $value_path ) {
+			$value = static::get_property_value( $styles, $value_path );
+
+			// Look up protected properties, keyed by value path.
+			// Skip protected properties that are explicitly set to `null`.
+			if ( is_array( $value_path ) ) {
+				$path_string = implode( '.', $value_path );
+				if (
+					array_key_exists( $path_string, static::PROTECTED_PROPERTIES ) &&
+					_wp_array_get( $settings, static::PROTECTED_PROPERTIES[ $path_string ], null ) === null
+				) {
+					continue;
+				}
+			}
+
+			// Skip if empty and not "0" or value represents array of longhand values.
+			$has_missing_value = empty( $value ) && ! is_numeric( $value );
+			if ( $has_missing_value || is_array( $value ) ) {
+				continue;
+			}
+
+			$declaration = array(
+				'name'  => $css_property,
+				'value' => $value,
+			);
+
+			if ( isset( static::LAYOUT_STYLES[ $css_property ] ) ) {
+				$declaration['selectors'] = static::LAYOUT_STYLES[ $css_property ];
```

Comment:
Nice trick!  I really like how the output CSS works.

I'd be interested to learn whether layout could be abstracted from this class somehow. I'm only mentioning since layout is growing and probably will grow more tendrils in the form of exceptions and further rules.

Reply:
Absolutely! Same here — my goal was to hack the code in wherever I could find to get this working. Now that it seems to mostly be working, the next step is to find a better and more logical home for the code changes 🙂

Let me know if you have any strong ideas on where anything should live!